rfcs/2023-05-03-data-volume-metrics.md

```diff
@@ -0,0 +1,238 @@
+
+The current mechanism to expire metrics is to check if a handle to the given
+metric is being held. If it isn't, and nothing has updated that metric in
+the last cycle - the metric is dropped. If a metric is dropped, the next time
+that event is emitted with those tags, the count starts at zero again.
+
+We will need to introduce a registered event caching layer that will register
+and cache new events keyed on the tags that are sent to it.
+
+Currently a registered metrics is stored in a `Registered<EventSent>`.
+
+We will need a new struct that can wrap this that will be generic over a tuple of
+the tags for each event and the event - eg. `Cached<(String, String), EventSent>`.
+This struct will maintain a BTreeMap of tags -> `Registered`. Since this will
+need to be shared across threads, the cache will need to be stored in an `RwLock`.
+
+In pseudo rust:
+
+```rust
+struct Cached<Tags, Event> {
+  cache: Arc<RwLock<BTreemap<Tags, Registered<Event>>>,
+  register: Fn(Tags) -> Registered<Event>,
+}
+
+impl<Tags, Event> Cached<Tags, Event> {
+  fn emit(&mut self, tags: Tags, value: Event) -> {
+    if Some(event) = self.cache.get(tags) {
+      event.emit(value);
+    } else {
+      let event = self.register(tags);
+      event.emit(value);
+      self.cache.insert(tags, event);
+    }
+  }
+}
+```
+
+## Rationale
+
+The ability to visualize data flowing through Vector will allow users to ascertain
+the effectiveness of the current use of Vector. This will enable users to
+optimise their configurations to make the best use of Vector's features.
+
+## Drawbacks
+
+The additional tags being added to the metrics will increase the cardinality of
+those metrics if they are enabled.
+
+## Prior Art
+
+
+## Alternatives
+
+We could use an alternative metric instead of estimated JSON size.
+
+- *Network bytes* This provides a more accurate picture of the actual data being received
+  and sent by Vector, but will regularly produce different sizes for an incoming event
+  to an outgoing event.
+- *In memory size* The size of the event as held in memory. This may be more accurate in
+  determining the amount of memory Vector will be utilizing at any time, will often be
+  less accurate compared to the data being sent and received which is often JSON.
+
+## Outstanding Questions
+
+## Plan Of Attack
+
+Incremental steps to execute this change. These will be converted to issues after the RFC is approved:
+
+- [ ] Add the `source` field to the Event metadata to indicate the source the event has come from.
+- [ ] Update the Volume event metrics to take a `JsonSize` value. Use the compiler to ensure all metrics
+      emitted use this. The `EstimatedJsonEncodedSizeOf` trait will be updated return a `JsonSize`.
+- [ ] Add the Service meaning. Update any sources that potentially create a service to point the meaning
+      to the relevant field.
+- [ ] Introduce an event caching layer that caches registered events based on the tags sent to it.
```

Comment:
I'd like to add benchmarking as an explicit part of the plan of action here. I think it'll be important to understand the performance impact.

Reply:
Have added.